### ISSUE TITLE
Performance problem when chaining calls on Enumerables of Eithers (e.g. BindT and MapT)

### ISSUE THREAD

**Issue by chrisbrook83** [louthy/language-ext#571]
I am using BindT and MapT to transform an enumerable of Eithers.

Using the 'T' operators to manipulate the IEnumerable directly (as opposed to using Map and then Bind/Map) is slower individually, but is non-linearly worse if you chain the operators together.

As an example:

```
[Test]
public void BindT_MapT_Slow()
{
    Either<bool, string>[] input = Enumerable.Repeat("1", 100000).Select(Prelude.Right<bool, string>).ToArray();

    // Fast: Map-Bind and Map-Map 
    var sw = Stopwatch.StartNew();
    var fastResult = input
        .Map(either => either.Bind(s => TryParseInt(s)))
        .Map(either => either.Map<string>(i => i.ToString()))
        .ToArray();
    Console.WriteLine($"Bind / Map: {sw.ElapsedMilliseconds}ms");
    
    // Slow: BindT and MapT
    sw.Restart();
    var slowResult = input
        .BindT(s => TryParseInt(s))
        .MapT(i => i.ToString())
        .ToArray();
    Console.WriteLine($"BindT / MapT: {sw.ElapsedMilliseconds}ms");

    /*** Local Func ***/
    Either<bool, int> TryParseInt(string s)
    {
        if (int.TryParse(s, out int i))
            return i;
        else
            return false;
    }
}
```

Results:
```
Bind / Map: 151ms
BindT / MapT: 45046ms
```

I'm running v3.1.15 on .NET Core 2.1

It doesn't seem to matter whether you are using MapT or BindT. For example, two chained MapT calls are similarly slow as the MapT & BindT as in the example above.

Furthermore, if you add additional transforms, the performance gets progressively worse:

```
// This takes 144,027ms on 100000 items
var verySlowResult = input
        .MapT(i => i.ToString())
        .MapT(i => i.ToString())
        .MapT(i => i.ToString())
        .ToArray();
```

PS: On the plus-side, at least we don't see any stack overflow issues running the latest code on .NET Core though! ;)
(re https://github.com/louthy/language-ext/issues/569)

**Comment by bender2k14** [comment]
I get a similar result for this test.

**Comment by louthy** [comment]
I would expect a difference in speed, because there will be lots of single item enumerables created.  But the difference is quite extreme.  It would require manual building of the transformer functions (for `BindT`) which would need updates to the T4 templates.  So, probably not going to happen super quick.

**Comment by louthy** [comment]
I have just run this with my Enumerable Bind and Concat performance improvements (released in 3.1.16) and the timing for both are the same.

**Comment by louthy** [comment]
If you're interested, [this is the code that got the gains](https://github.com/louthy/language-ext/blob/master/LanguageExt.Core/Utility/EnumerableOptimal.cs)
